File: recipes/claude/26-remote-mcp-servers-and-tunnels/typescript/src/remote_server.ts
import express, { Request, Response, NextFunction } from "express";
import cors from "cors";
import crypto from "crypto";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { SSEServerTransport } from "@modelcontextprotocol/sdk/server/sse.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";

const PORT = parseInt(process.env.MCP_PORT || "8000", 10);
const AUTH_TOKEN = process.env.MCP_AUTH_TOKEN || "default-insecure-dev-token-change-in-production";

// Initialize core MCP Server instance
const mcpServer = new Server(
  {
    name: "ProductionRemoteOpsTS",
    version: "1.0.0",
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

// Register Available Tools
mcpServer.setRequestHandler(ListToolsRequestSchema, async () => {
  return {
    tools: [
      {
        name: "query_system_metrics",
        description: "Retrieve operational health metrics from the remote server environment.",
        inputSchema: {
          type: "object",
          properties: {},
        },
      },
      {
        name: "execute_database_diagnostics",
        description: "Run read-only database cluster connection diagnostics.",
        inputSchema: {
          type: "object",
          properties: {
            cluster_id: { type: "string", description: "Target database cluster identifier" },
          },
          required: ["cluster_id"],
        },
      },
    ],
  };
});

// Handle Tool Execution Requests
mcpServer.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  if (name === "query_system_metrics") {
    return {
      content: [
        {
          type: "text",
          text: JSON.stringify({
            status: "healthy",
            load_average: [0.12, 0.08, 0.05],
            uptime_seconds: 86400,
            active_tenants: 4,
            protocol_version: "2024-11-05",
          }),
        },
      ],
    };
  }

  if (name === "execute_database_diagnostics") {
    const clusterId = args?.cluster_id as string;
    return {
      content: [
        {
          type: "text",
          text: JSON.stringify({
            cluster_id: clusterId,
            connection_pool: { active: 12, idle: 28, max: 50 },
            replication_lag_ms: 1.4,
            status: "synchronized",
          }),
        },
      ],
    };
  }

  throw new Error(`Tool not found: ${name}`);
});

// Track active SSE transports by sessionId
const activeTransports = new Map<string, SSEServerTransport>();

// Express Application Setup
const app = express();
app.use(cors({ origin: "*", exposedHeaders: ["Mcp-Session-Id"] }));

// Constant-time token verification middleware
const authenticateBearer = (req: Request, res: Response, next: NextFunction): void => {
  const authHeader = req.headers.authorization;
  if (!authHeader || !authHeader.startsWith("Bearer ")) {
    res.status(401).json({ error: "Unauthorized", message: "Valid Bearer token required" });
    return;
  }

  const token = authHeader.slice("Bearer ".length).trim();
  const tokenBuffer = Buffer.from(token);
  const secretBuffer = Buffer.from(AUTH_TOKEN);

  if (tokenBuffer.length !== secretBuffer.length || !crypto.timingSafeEqual(tokenBuffer, secretBuffer)) {
    res.status(401).json({ error: "Unauthorized", message: "Invalid Bearer token" });
    return;
  }

  next();
};

// Health Check Probe
app.get("/health", (_req: Request, res: Response) => {
  res.json({
    status: "online",
    service: "remote-mcp-server-ts",
    active_sessions: activeTransports.size,
  });
});

// SSE Handshake Endpoint
app.get("/sse", authenticateBearer, async (req: Request, res: Response) => {
  const transport = new SSEServerTransport("/messages", res);
  const sessionId = transport.sessionId;
  activeTransports.set(sessionId, transport);

  res.setHeader("Mcp-Session-Id", sessionId);

  req.on("close", () => {
    activeTransports.delete(sessionId);
  });

  await mcpServer.connect(transport);
});

// JSON-RPC Post Message Endpoint
app.post("/messages", authenticateBearer, async (req: Request, res: Response) => {
  const sessionId = (req.query.sessionId as string) || (req.headers["mcp-session-id"] as string);

  if (!sessionId || !activeTransports.has(sessionId)) {
    res.status(400).json({ error: "Bad Request", message: "Invalid or expired sessionId" });
    return;
  }

  const transport = activeTransports.get(sessionId)!;
  await transport.handlePostMessage(req, res);
});

app.listen(PORT, "0.0.0.0", () => {
  console.log(`Production Remote MCP Server (TS) listening on http://0.0.0.0:${PORT}`);
});

File: recipes/claude/26-remote-mcp-servers-and-tunnels/python/remote_server.py
#!/usr/bin/env python3
"""
Remote Model Context Protocol (MCP) Server with Bearer Authentication & SSE.
Designed for secure reverse proxy deployment behind Nginx, Cloudflare Tunnels, or Tailscale Funnel.
"""

import os
import hmac
import secrets
from typing import Dict
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import JSONResponse, Response
from starlette.middleware import Middleware
from starlette.middleware.cors import CORSMiddleware
from starlette.routing import Route
import uvicorn
from mcp.server.fastmcp import FastMCP
from mcp.server.sse import SseServerTransport

# Configuration from environment
AUTH_TOKEN = os.getenv("MCP_AUTH_TOKEN", "default-insecure-dev-token-change-in-production")
PORT = int(os.getenv("MCP_PORT", "8000"))
HOST = os.getenv("MCP_HOST", "0.0.0.0")

# Initialize FastMCP application instance
mcp = FastMCP("ProductionRemoteOps")

@mcp.tool()
def query_system_metrics() -> dict:
    """Retrieve operational health metrics from the remote server environment."""
    return {
        "status": "healthy",
        "load_average": [0.12, 0.08, 0.05],
        "uptime_seconds": 86400,
        "active_tenants": 4,
        "protocol_version": "2024-11-05"
    }

@mcp.tool()
def execute_database_diagnostics(cluster_id: str) -> dict:
    """Run read-only database cluster connection diagnostics."""
    return {
        "cluster_id": cluster_id,
        "connection_pool": {"active": 12, "idle": 28, "max": 50},
        "replication_lag_ms": 1.4,
        "status": "synchronized"
    }

# Active SSE Transports indexed by Session ID
active_transports: Dict[str, SseServerTransport] = {}

def verify_token(authorization_header: str | None) -> bool:
    """Constant-time bearer token validation to prevent timing attacks."""
    if not authorization_header or not authorization_header.startswith("Bearer "):
        return False
    incoming_token = authorization_header[len("Bearer "):].strip()
    return secrets.compare_digest(incoming_token, AUTH_TOKEN)

async def handle_sse(request: Request) -> Response:
    """Handle SSE connection initiation and register session."""
    auth_header = request.headers.get("Authorization")
    if not verify_token(auth_header):
        return JSONResponse({"error": "Unauthorized", "message": "Valid Bearer token required"}, status_code=401)

    transport = SseServerTransport("/messages/")
    session_id = secrets.token_urlsafe(16)
    active_transports[session_id] = transport

    async with transport.connect_sse(request.scope, request.receive, request._send) as streams:
        await mcp._mcp_server.run(streams[0], streams[1], mcp._mcp_server.create_initialization_options())
    
    active_transports.pop(session_id, None)
    return Response(status_code=200)

async def handle_messages(request: Request) -> Response:
    """Handle incoming JSON-RPC client messages routed to active SSE session."""
    auth_header = request.headers.get("Authorization")
    if not verify_token(auth_header):
        return JSONResponse({"error": "Unauthorized", "message": "Valid Bearer token required"}, status_code=401)

    session_id = request.query_params.get("sessionId") or request.headers.get("Mcp-Session-Id")
    if not session_id or session_id not in active_transports:
        return JSONResponse({"error": "Bad Request", "message": "Invalid or missing sessionId"}, status_code=400)

    transport = active_transports[session_id]
    await transport.handle_post_message(request.scope, request.receive, request._send)
    return Response(status_code=202)

async def handle_health(request: Request) -> Response:
    """Health check endpoint for reverse proxy probes."""
    return JSONResponse({
        "status": "online",
        "service": "remote-mcp-server",
        "active_sessions": len(active_transports)
    })

routes = [
    Route("/sse", endpoint=handle_sse, methods=["GET"]),
    Route("/messages/", endpoint=handle_messages, methods=["POST"]),
    Route("/health", endpoint=handle_health, methods=["GET"])
]

middleware = [
    Middleware(
        CORSMiddleware,
        allow_origins=["*"],
        allow_methods=["GET", "POST", "OPTIONS"],
        allow_headers=["*"],
        expose_headers=["Mcp-Session-Id"]
    )
]

app = Starlette(routes=routes, middleware=middleware)

if __name__ == "__main__":
    uvicorn.run(app, host=HOST, port=PORT, log_level="info")
